### ISSUE TITLE
Enable Provenance on iPhone to act as Game Controller to Provenance on AppleTV

### ISSUE THREAD

**Issue by zonabi** [jasarien/Provenance#136]
As an iOS Gamer, it would be great to be able to use the iPhone Provenance App as a Game Controller to the Provenance AppleTV App. 

The Provenance iOS app could have a new View that is a simple Game Controller that sends messages to the Provenance tvOS app.

**Comment by jpete24** [comment]
This actually would be a pretty neat idea. Not sure how well it could be implemented, but not a bad solution to the controller problem.

**Comment by damian-pf** [comment]
+1 this is a great idea!

**Comment by tkunstek** [comment]
I saw this in another emulator for iOS a number of years ago here: https://github.com/rileytestut/SNES4iOS

Effectively you could use an iPhone as a remote control for the iPad.  Demo'ed great to the class.  However, this emulator broke in iOS 8 and has not been supported since.

**Comment by SamWarmuth** [comment]
:+1:

**Comment by yusefnapora** [comment]
I actually wrote the SNES-HD app that the above link was branched off of long ago :)  I ended up re-writing the remote gamepad stuff so that the iPad app serves up a mobile-friendly webpage that would render an SVG gamepad and use javascript for interaction.

Life got in the way, and I never finished my rewrite, but since I'm getting an Apple TV in a few days, I've been thinking about digging that code out of the bin and seeing if it could be useful here.  There's an embedded web server in Provenance anyway... 

I'll see what I can come up with :)

**Comment by jasarien** [comment]
No JavaScript plz ;)

**Comment by yusefnapora** [comment]
JavaScript is messy, no doubt :)

The main advantage is that you don't need to install anything on the phone, which makes it easier for friends to join in since they don't need to connect to Xcode & provision their devices, etc.  It also means you don't need to worry about versioning if you ever change the wire protocol, since you control both ends.

But if that's not a big concern, I'd rather do native code as well.

**Comment by jasarien** [comment]
My main concern is latency from the web view, over wifi, through the http server, translated into button presses sent to the emulator - I can't see that working at all for some games that need immediate feedback for button presses. Even the native on-screen controls aren't as latency free as I'd like, and the MFi controllers aren't the best either.

**Comment by XeresRazor** [comment]
We'd be better off supporting something like Joypad.Joypad Game Controller by Joypad
https://appsto.re/us/EhP1E.i
I'll take a look at their sdk today.

**Comment by zonabi** [comment]
bluetooth, y'all! no wifi. yes @XeresRazor exactly like the Joypad app/controller

**Comment by davidmuzi** [comment]
You can setup a socket connection between the devices, faster than BT and using a web server

**Comment by jasarien** [comment]
Agreed.

**Comment by alenofx** [comment]
I've tested the Joypad SDK on Provenance just for fun, creating the required GCController related subclasses (very similar to the iCade implementation). With an iPhone as controller and an iPad as player it works just fine.
Pro: easy integration, nice UI, good latency*.
Cons: *bad latency after a few minutes of gameplay, old and abandoned closed source SDK, no arm64 support.

Anyway, I agree with a custom direct connection solution using sockets.

**Comment by alenofx** [comment]
An interesting update: here is a new (Swift) framework to create virtual game controllers based on NSNetService: [VirtualGameController](https://github.com/robreuss/VirtualGameController).
Haven't tested it yet on Provenance.

**Comment by Oggymator** [comment]
that would be great or is there any game controller app that emulate a game controller on the iPhone to use it in more games?

**Comment by Oggymator** [comment]
is it possible to use this in any way?
https://github.com/robreuss/VirtualGameController

**Comment by rohapa** [comment]
Apple just released tvos 9.1 for Apple TV.
Allows Remote Control with iphone now.

Don't know if that helps, but yeeehhaa!
http://9to5mac.com/2015/12/08/tvos-9-1/

**Comment by graphicbryan** [comment]
Release of 9.1 for tvOS makes responsiveness of both the SteelSeries Nimbus & Horipad Ultimate both nearly flawless for MFi controller input.

**Comment by jasarien** [comment]
wait, really?

**Comment by graphicbryan** [comment]
I tested from GBA, NES, Genesis, SNES, and played all the way up to Venom in Star Fox, and felt just like it was 1993 all over again. 

Also there was no reason to re-compile Providence, the update to 9.1 did not remove anything other than my developer menu in the Settings app, not a big deal easy to re-enable with Xcode 7.2 (also updated yesterday).

**Comment by jasarien** [comment]
Cool, lets take this over to https://github.com/jasarien/Provenance/issues/157 and discuss it there.

**Comment by JakeBondo** [comment]
I would love to see this added. 👍🏼

**Comment by sevdestruct** [comment]
What's status on this? Has anyone tried this before: [Apple TV Remote App doubles as game controller](http://www.ilounge.com/index.php/tips/comments/using-your-iphone-as-a-game-controller-for-your-apple-tv)

Do we really need Provenance app to do this? Have them speak app to app, iPhone to aTV? or does the Apple TV Remote app for iPhone suffice? or is there another 3rd party app out there that already turns your phone into virtual MFi controller. If so, I'd like to close this one. We have full MFi controller support already making this just a novelty (who wants to play on the tv without a true haptic physical muscle-memory buttons?), and @JoeMatt  is working on a great solution for getting full featured controllers (like PS3, etc) to talk to the app through another means. 

Can we close this?

**Comment by JoeMatt** [comment]
This is possible with VirtualGameController framework. I did some testing a few weeks ago with their latest release. Seems to work. I might use it in my other code to make bluetooth controllers looks like MFi.

**Comment by Vercantez** [comment]
@sevdestruct There doesn't seem to be any way to turn iPhones into virtual MFi controllers because iPhone does not support the HID bluetooth profile needed. At least not through CoreBluetooth.
@JoeMatt VGC looks great, however it would need source changes to Providence. It looks like it is designed to be a drop-in replacement, so it may not be a lot of changes. I'll try to make a fork that implements VGC when I get some time.

**Comment by JoeMatt** [comment]
Yeah VGC is the way to go. They have iCade support too, so instead of using our own iCade classes we should use theirs and add any missing iCade profiles they might not have and push those changes back upstream. The best way to do that is to fork them on GH and use a git submodule, then you can make a PR against the fork, and if they include the changes in their mainline, we just change the submodule URL to their repo. It's how we're doing cores now too in 1.5, transitioning as I update them.

**Comment by mac-user669** [comment]
Any updates? Has anyone successfully used their iPhone to control Provenance on an Apple TV?

**Comment by JoeMatt** [comment]
Controller code is going under a refactor. That's stage 1.

Stage 2 would be adding in VirtualGameController (if it's still under active dev I haven't checked, or something similar).

Unless making a contribution, please don't use GitHub issues for "me too" or "any updates?" posts.
If there were updates, the git commits would be tagged with this ticket and would show up here. Any discussion can be directed to Discord, linked on the main page.

**Comment by mac-user669** [comment]
Ok, thanks. Alright, I'll try not to clutter this up. Thanks!

**Comment by jbellanca** [comment]
This would be phenomenal for upcoming Intellivision (and Colecovision) emulation, since they have controllers with keypads and overlays that can't be replicated with a modern controller. Kind of like how Nintellivision for the DS handles it.

**Comment by JoeMatt** [comment]
@claude we should consider this at some point, really 2 things,

1. we could just add DSU support, and a DSU client in app (i did it in icube already Provenance-EMU/iCube)
2. the idea @jbellanca had for systems that have way too many buttons, like those with number pads (various atari systems, coleceovision, vectrex, keyboard and mouse for dos, wolf 3d etc) where the phone could be used as an extended controller for another device, ipad, tvos app or while doing the upcoming #1641 #2684 #2687 airplay mode (which also has skin suppport for airplay skinned controllers).

DSU seems not too hard, especially since i wrote swiftui code for it already in icube's repo which is open-source

make epic/task tickets, don't do the coding yet just the research and tickets
